File: src/main/java/destiny/astrology/classical/rules/debilities/Partile_Oppo_Mars_Saturn.java
/**
 * @author smallufo
 * Created on 2007/12/31 at 上午 3:38:12
 */
package destiny.astrology.classical.rules.debilities;

import destiny.astrology.*;
import org.jetbrains.annotations.NotNull;
import org.jooq.lambda.tuple.Tuple;
import org.jooq.lambda.tuple.Tuple2;

import java.util.Optional;

import static java.util.Optional.empty;

/** Partile opposite Mars or Saturn. */
public final class Partile_Oppo_Mars_Saturn extends Rule {

  public Partile_Oppo_Mars_Saturn() {
  }

  @Override
  protected Optional<Tuple2<String, Object[]>> getResult(Planet planet, @NotNull Horoscope h) {
    return h.getPositionOptional(planet).map(Position::getLng).flatMap(planetDegree ->
      h.getPositionOptional(Planet.MARS).map(Position::getLng).flatMap(marsDeg ->
        h.getPositionOptional(Planet.SATURN).map(Position::getLng).flatMap(saturnDeg -> {
          if (planet != Planet.MARS && AspectEffectiveModern.isEffective(planetDegree, marsDeg, Aspect.OPPOSITION, 1.0)) {
            logger.debug("{} 與 {} 形成 {}" , planet , Planet.MARS , Aspect.OPPOSITION);
            return Optional.of(Tuple.tuple("comment", new Object[]{planet, Planet.MARS, Aspect.OPPOSITION}));
          }
          else if (planet != Planet.SATURN && AspectEffectiveModern.isEffective(planetDegree, saturnDeg, Aspect.OPPOSITION, 1.0)) {
            logger.debug("{} 與 {} 形成 {}" , planet , Planet.SATURN , Aspect.OPPOSITION);
            return Optional.of(Tuple.tuple("comment", new Object[]{planet, Planet.SATURN, Aspect.OPPOSITION}));
          }
          return empty();
        })
      )
    );

  }

}

File: src/main/java/destiny/astrology/classical/rules/debilities/Partile_Oppo_Mars_Saturn.kt
/**
 * @author smallufo
 * Created on 2007/12/31 at 上午 3:38:12
 */
package destiny.astrology.classical.rules.debilities

import destiny.astrology.Aspect.OPPOSITION
import destiny.astrology.AspectEffectiveModern
import destiny.astrology.IHoroscopeModel
import destiny.astrology.Planet
import destiny.astrology.Planet.MARS
import destiny.astrology.Planet.SATURN

/** Partile opposite Mars or Saturn.  */
class Partile_Oppo_Mars_Saturn : DebilityRule() {

  private val aspect = OPPOSITION

  override fun getResult(planet: Planet, h: IHoroscopeModel): Pair<String, Array<Any>>? {
    val planetDeg = h.getPosition(planet)?.lng
    val marsDeg = h.getPosition(MARS)?.lng
    val saturnDeg = h.getPosition(SATURN)?.lng

    return planetDeg?.let {
      val marResult = marsDeg?.takeIf {
        planet !== MARS && AspectEffectiveModern.isEffective(planetDeg, marsDeg, aspect, 1.0)
      }?.let { "comment" to arrayOf(planet, MARS, aspect) }

      val satResult = saturnDeg?.takeIf {
        planet != SATURN && AspectEffectiveModern.isEffective(planetDeg, saturnDeg, aspect, 1.0)
      }?.let { "comment" to arrayOf(planet, SATURN, aspect) }

      marResult ?: satResult
    }
  }
}
